.github/actions/javascript/authorChecklist/newComponentCategory.ts

```diff
@@ -0,0 +1,104 @@
+    'I verified that all JSX used for rendering exists in the render method',
+    'I verified that each component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions',
+];
+
+function detectReactComponent(code: string, filename: string): boolean | undefined {
+    if (!code) {
+        console.error('failed to get code from a filename', code, filename);
+        return;
+    }
+    const ast = parse(code, {
+        sourceType: 'module',
+        plugins: ['jsx'], // enable jsx plugin
+    });
+
+    let isReactComponent = false;
+
+    traverse(ast, {
+        enter(path) {
+            if (isReactComponent) {
+                return;
+            }
+            if (path.isFunctionDeclaration() || path.isArrowFunctionExpression() || path.isFunctionExpression()) {
+                path.traverse({
+                    // eslint-disable-next-line @typescript-eslint/naming-convention
+                    JSXElement() {
+                        isReactComponent = true;
+                        path.stop();
+                    },
+                });
+            }
+        },
+        // eslint-disable-next-line @typescript-eslint/naming-convention
+        ClassDeclaration(path) {
+            const {superClass} = path.node as unknown as SuperClassType;
+            if (superClass && ((superClass.object && superClass.object.name === 'React' && superClass.property.name === 'Component') || superClass.name === 'Component')) {
+                isReactComponent = true;
+                path.stop();
+            }
+        },
+    });
+
+    return isReactComponent;
+}
+
+function nodeBase64ToUtf8(data: string) {
+    return Buffer.from(data, 'base64').toString('utf-8');
+}
+
+async function detectReactComponentInFile(filename: string): Promise<boolean | undefined> {
+    const params = {
+        owner: CONST.GITHUB_OWNER,
+        repo: CONST.APP_REPO,
+        path: filename,
+        ref: github.context.payload.pull_request?.head.ref,
+    };
+    try {
+        const {data} = await GithubUtils.octokit.repos.getContent(params);
+        const content = 'content' in data ? nodeBase64ToUtf8(data.content || '') : data;
+        return detectReactComponent(content, filename);
+    } catch (error) {
+        console.error('An unknown error occurred with the GitHub API: ', error, params);
+    }
+}
+
+function filterFiles({filename, status}: {filename: string; status: string}): boolean {
+    if (status !== 'added') {
+        return false;
+    }
+    return filename.endsWith('.js') || filename.endsWith('.jsx') || filename.endsWith('.ts') || filename.endsWith('.tsx');
+}
+
+async function detect(changedFiles: Array<{filename: string; status: string}>): Promise<boolean> {
+    const filteredFiles = changedFiles.filter(filterFiles);
+    for (const file of filteredFiles) {
+        const result = await detectReactComponentInFile(file.filename);
```

Comment:
Similar story - there can be many changed files in a PR, so we really want to avoid using `await` in a loop here. It took some Promise-fu, but I found what I believe to be a clever way to parallelize this. It uses a custom function inspired by `Promise.race` called `promiseRaceTo`:

```ts
/**
 * Like Promise.race, except it resolves with the first promise that resolves to the desired value.
 * If no promise resolves to the desired value, it rejects.
 */
export default function promiseRaceTo<T>(promises: Array<Promise<unknown>>, desiredValue: T): Promise<T> {
    return new Promise((resolve, reject) => {
        for (const p of promises) {
            Promise.resolve(p)
                .then((res) => {
                    if (res !== desiredValue) {
                        return;
                    }
                    resolve(res as T);
                })
                .catch(() => {});
        }
        Promise.allSettled(promises).then(() => reject());
    });
}
```

Then you can use it like this:

```ts
async function detect(changedFiles: Array<{filename: string; status: string}>): Promise<boolean> {
    const filteredFiles = changedFiles.filter(({filename, status}) => status === 'added' && (filename.endsWith('.js') || filename.endsWith('.ts') || filename.endsWith('.tsx')));
    try {
        // Check all files that may contain a React component in parallel
        // and resolve as soon as the first one is found
        await promiseRaceTo(
            filteredFiles.map(({filename}) => detectReactComponentInFile(filename)),
            true,
        );
        return true;
    } catch (err) {
        // All files have been checked and none contain a React component
        return false;
    }
}
```

Reply:
Great job. I've been trying to use `Promise.race` but as you know it won't work in this case.